core/dnsserver/zdirectives.go

```diff
@@ -46,6 +46,7 @@ var Directives = []string{
 	"clouddns",
 	"k8s_external",
 	"kubernetes",
+	"kubernetescrd",
```

Comment:
Another difference between routing queries through separate serverblocks vs single point will surface in metrics, since server scope metrics have a zone label of the enclosing serverblock.  When routing queries through one point, the resulting metrics would all be labeled with the same zone (the zone of the serverblock enclosing the plugin).

Reply:
> If the plugin is expanded to expose other functions via CRDs, then picking a single position in the plugin chain that is appropriate for all functions may prove difficult/impossible.

Regarding the above: This may be solved by instead of having the crd-monitoring plugin serving DNS requests, the other plugins did.  For example, if forward could handle multiple zones, learn of added/deleted zones from the crd-monitoring plugin while running, and start/stop serving those zones.  In this way, dns requests would be served by the forward plugin itself.  Other plugins could be adapted similarly, e.g. hosts, for crd controlled A/AAAA records.  The hosts and forward plugins require different positions in the plugin chain (hosts near the top, and forward near the bottom).

Reply:
> The position of this plugin in the chain is driven by the DNS function performed, which for now is just forwarding. If the plugin is expanded to expose other functions via CRDs, then picking a single position in the plugin chain that is appropriate for all functions may prove difficult/impossible. IOW, this plugin may need to remain specific to stub domain forwarding.

Positioning in the chain is an interesting point I didn't realize. I was optimistically aiming for this plugin to encapsulate other functions via CRDs, but it does seem difficult now. Perhaps it is best to rename this plugin if you think it is unlikely we can encapsulate other functionality into this plugin (maybe `forwardingcrd` or something)? It does feel like it may get unwieldy if there becomes a proliferation of plugins to implement more functionality via CRDs.

> We might want to order this immediately before forward. In its current position, it executes before file, whereas forward normally executes after file. I think we should try to make the behavior be as close as possible to using the forward plugin.

Makes sense, and I can make this change.

> However, the typical way of configuring a stub domain is to define a serverblock for the zone. A separate serverblock creates a plugin chain for that zone, allowing for different configurations per stubdomain (e.g. cache settings). Routing all the zones from a single point isn't quite as flexible. There is currently no way for plugins to "create" new server blocks. I think it would be possible to modify core/dnsserver to enable that, but it could get complicated.

Yeah, the current implementation is not very flexible in this regard. I like the idea of this plugin "creating" new server blocks, but it does seem very complex and I did not want to dig too deep into modifying core/dnsserver code on the initial implementation. I could see the plugin being refactored to use such functionality and perhaps the CRD at that point could allow a user to specify what plugins they want to enable and those could be translated into server blocks. Perhaps creating new server blocks would also helps with the plugin chain ordering problem and the metrics problem?

> Regarding the above: This may be solved by instead of having the crd-monitoring plugin serving DNS requests, the other plugins did. For example, if forward could handle multiple zones, learn of added/deleted zones from the crd-monitoring plugin while running, and start/stop serving those zones. In this way, dns requests would be served by the forward plugin itself. Other plugins could be adapted similarly, e.g. hosts, for crd controlled A/AAAA records. The hosts and forward plugins require different positions in the plugin chain (hosts near the top, and forward near the bottom).

That makes sense, but one thing I'm unclear on, do instances of plugins just exist that I can modify (assuming we gave that plugin the functionality to be externally modified), or would the crd-monitoring plugin still have to instantiate instances of the plugin (and would that mean it isn't part of the proper plugin chain)?

Reply:
> Perhaps creating new server blocks would also helps with the plugin chain ordering problem and the metrics problem?

Yes - it would essentially mimic the current way stub zones are configured.  So, metrics would get labeled in the correct zone.  Plugin chain ordering would not matter, since each zone gets its own chain.

I've been looking into how this could be implemented, and what it looks like.  After a few hours fiddling with it, it can confirm, it _is_ complicated - and requires some (perhaps unacceptable) hacks.  ATM, I'm thinking that my tinkering wont produce something feasible, and we'll just document the differences in behavior between CRD stub zones and manually defined ones.

> That makes sense, but one thing I'm unclear on, do instances of plugins just exist that I can modify (assuming we gave that plugin the functionality to be externally modified), or would the crd-monitoring plugin still have to instantiate instances of the plugin (and would that mean it isn't part of the proper plugin chain)?

I was thinking we assume they exist.  Dynamically inserting another plugin into the chain while maintaining the correct execution order would almost impossible I think.  Assuming the plugin exists means we'd probably want the dns-serving plugins to be declared in a blank slate state (to later be configured via the crd interface). But in the CRD stub zone forward case, an instance of forward typically pre-exists (the default upstream).

The crd plugin would not sit in the plugin chain. It would not have a `ServeDNS()` function defined (wont implement a plugin Handler), and it would not be added to the Config in setup().  There are examples of this kind of "configuration" type plugin, e.g. *sign*

Another approach entirely could be to create "crd" versions of each plugin we want to CRD-ify... i.e. *forward-crd*
, *hosts-crd*, ... and package some of the shared crd monitoring components to reduce code duplication. These crd plugins could be placed next to (just before, just after?) their non-crd counterparts. *forward-crd* essentially akin to the implementation in this PR.  This minimizes ordering angst, but i think there may still be a little WRT the before/after question (i.e. which one "wins" in the event of zone overlap).

Reply:
Any preference on which implementation is preferable? Not sure if I lean towards any particular implementation since they both seem to have their own tradeoffs. Configuring a separate plugin means you have to add both plugins to the chain, which could lead to misconfiguration, but solves the ordering issue. CRD-ifying plugins solves adding both plugins to the chain, but there is now an ordering issue with their non-crd counterparts.

Reply:
> Any preference on which implementation is preferable?

I like the "crd plugin configures other plugins" solution because it mimics the behavior of having had configured those plugins via the Corefile, and it results in fewer plugins down the road.
However, I like the "forward-crd plugin serves dns itself via an instance of forward" because it would enable other plugins to leverage the forwarding code of the forward plugin.

I'm leaning toward "forward-crd plugin serves dns itself via an instance of forward" ... since i'm on the fence, and you've more or less already done it this way.

Reply:
Sounds good. I've done a quick rename from `kubernetescrd` to `forwardcrd`. I think that is all I might need to do at this point, but let me know if I've missed anything.